### PR TITLE
GH-25: Add PhoneDialer API

### PR DESCRIPTION
**Pull request by F1nZeR**
#25 

Will be updated according to related discussions.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [ ] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense
- [x] Has Samples

### CODE REVIEW

**Comment by F1nZeR** [comment]
I've added temporary exception classes.

File: Caboodle/PhoneDialer/PhoneDialer.android.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace Microsoft.Caboodle
+{
+	public static partial class PhoneDialer
+	{
+		public static bool IsSupported
+		{
+			get
+			{
+				var packageManager = Application.Context.PackageManager;
```
Comment (jamesmontemagno) [inline Caboodle/PhoneDialer/PhoneDialer.android.cs]: My thoughts here are to also check telephony is actually on the device.

File: Caboodle/PhoneDialer/PhoneDialer.ios.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace Microsoft.Caboodle
+{
+	public static partial class PhoneDialer
+	{
+		public static bool IsSupported
+		{
+			get
+			{
+				var nsUrl = CreateNsUrl("0000000000");
```
Comment (jamesmontemagno) [inline Caboodle/PhoneDialer/PhoneDialer.ios.cs]: shouldn't this append with: tel:

File: Caboodle/PhoneDialer/PhoneDialer.ios.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+				var canCall = UIApplication.SharedApplication.CanOpenUrl(nsUrl);
+				
+				if (canCall)
+				{
+					using (var netInfo = new CTTelephonyNetworkInfo())
+					{
+						var mnc = netInfo.SubscriberCellularProvider?.MobileNetworkCode;
+						return !string.IsNullOrEmpty(mnc) && mnc != "65535"; // 65535 stands for NoNetwordProvider
```
Comment (jamesmontemagno) [inline Caboodle/PhoneDialer/PhoneDialer.ios.cs]: don't need to check against null here.

File: Caboodle/PhoneDialer/PhoneDialer.ios.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+				}
+				
+				return false;
+			}
+		}
+
+		public static void Open(string number)
+		{
+			if (string.IsNullOrWhiteSpace(number))
+			{
```
Comment (jamesmontemagno) [inline Caboodle/PhoneDialer/PhoneDialer.ios.cs]: @mattleibow are we standardizing on need or not needing { around a single line if?

File: Caboodle/PhoneDialer/PhoneDialer.ios.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+			if (!IsSupported)
+			{
+				throw new CapabilityNotSupportedException();
+			}
+			
+			var nsUrl = CreateNsUrl(number);
+			UIApplication.SharedApplication.OpenUrl(nsUrl);
+		}
+
+		private NSUrl CreateNsUrl(string number)
```
Comment (jamesmontemagno) [inline Caboodle/PhoneDialer/PhoneDialer.ios.cs]: Should use expression bodied method here.

File: Caboodle/PhoneDialer/PhoneDialer.uwp.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+using Windows.ApplicationModel.Calls;
+
+namespace Microsoft.Caboodle
+{
+	public static partial class PhoneDialer
+	{
+		public static bool IsSupported =>
+			 Windows.Foundation.Metadata.ApiInformation.IsTypePresent("Windows.ApplicationModel.Calls.PhoneCallManager");
+
+		public static void Open(string number)
```
Comment (jamesmontemagno) [inline Caboodle/PhoneDialer/PhoneDialer.uwp.cs]: Expression bodied Method here

File: Caboodle/PhoneDialer/PhoneDialer.ios.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+		public static bool IsSupported
+		{
+			get
+			{
+				var nsUrl = CreateNsUrl("0000000000");
+				var canCall = UIApplication.SharedApplication.CanOpenUrl(nsUrl);
+				
+				if (canCall)
+				{
+					using (var netInfo = new CTTelephonyNetworkInfo())
```
Comment (jamesmontemagno) [inline Caboodle/PhoneDialer/PhoneDialer.ios.cs]: What all is this check doing?

**Comment by jamesmontemagno** [comment]
The other question I have is what should the user pass in here as it looks like ANdroid does special formatting....

Do they need +1 for instance? does 555.555.5555 work?

File: Caboodle/PhoneDialer/PhoneDialer.ios.cs (hunk 4 above)
Comment (mattleibow) [reply Caboodle/PhoneDialer/PhoneDialer.ios.cs]: @jamesmontemagno I prefer always using braces for the simple reason that it avoids confusion due to the grouping. Also it makes changes easier - if we ever have to add another line, then we don't have to also add braces. But, this is a very opinionated thing. 

If we are following the .net foundation guides, then they say that we don't need it. So I guess we don't need them.
https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

Comment (mattleibow) [reply Caboodle/PhoneDialer/PhoneDialer.ios.cs]: Then again, we don't really follow most of them.

File: Caboodle/PhoneDialer/PhoneDialer.ios.cs (hunk 3 above)
Comment (F1nZeR) [reply Caboodle/PhoneDialer/PhoneDialer.ios.cs]: All these properties might be null.
https://developer.apple.com/documentation/coretelephony/cttelephonynetworkinfo/1616900-subscribercellularprovider
https://developer.apple.com/documentation/coretelephony/ctcarrier/1620324-mobilenetworkcode

File: Caboodle/PhoneDialer/PhoneDialer.ios.cs (hunk 7 above)
Comment (F1nZeR) [reply Caboodle/PhoneDialer/PhoneDialer.ios.cs]: we are checking for SIM-card & network availability

**Comment by mattleibow** [comment]
@F1nZeR to save you time merging and resolving conflicts, I did it here: https://github.com/F1nZeR/Caboodle/pull/1

NOTE: I left your files as tabs so that it (hopefully) won't break any local changes you are working on.

**Comment by jamesmontemagno** [comment]
We now have checks in place for StyleCop + Unit Tests + Samples ready to go.

Are you able to ensure everything pulls up from master and validates against StyleCop. Additionally can you add Unit Tests, Tests for Device Runners, and Samples.

**Comment by jamesmontemagno** [comment]
I think the only thing we are trying to figure out is if we want "IsSupported" or some enum there.

**Comment by jamesmontemagno** [comment]
Pulling into a branch so I can work on it

**Pull request merged**